### PR TITLE
fix(ui): secret field json diff visibility issue

### PR DESCRIPTION
**Pull request by IgorHorta**
## Context

Previously, when reviewing secret changes in the approval workflow, users could only see that a secret value had changed, but couldn't easily identify exactly what changed within the value. This made it difficult to review changes, especially for long strings or multiline values.

This PR implements an enhanced diff visualization for secret approvals that provides:
- **Line-by-line comparison** for multiline secret values with visual indicators (`+` for additions, `-` for deletions)
- **Word-level highlighting** within modified lines to pinpoint exactly which words/characters changed
- **Single-line diff support** with inline word highlighting (no `+/-` indicators for cleaner display)
- **Auto-scroll to first change** to help users quickly find what changed
- **Enhanced visual design** with darker tinted backgrounds (`#120808` for old, `#081208` for new) for better contrast
- **Proper text wrapping** for long strings to prevent overflow

The implementation uses a word-boundary-aware diff algorithm similar to GitHub's diff view, making it much easier to review and approve secret changes.

Related to: SECRETS-65

## Screenshots

https://github.com/user-attachments/assets/2cb17a12-81b1-4413-a6b3-0d2e2a907780

### Test Case 1: Single-Line Secret Diff
1. Navigate to a secret approval request
2. Find a secret where the value is a single line (no newlines) and has been changed
3. **Expected**: 
   - The changed words/characters are highlighted inline (red for old, green for new)
   - No `+/-` indicators or line containers
   - Text wraps if the value is very long

### Test Case 2: Multiline Secret Diff
1. Navigate to a secret approval request
2. Find a secret where the value contains newlines and has been changed
3. **Expected**:
   - Changed lines show `-` (red) for deletions and `+` (green) for additions
   - Modified lines show both old and new versions with word-level highlighting
   - Full lines are highlighted with the appropriate background color
   - Container automatically scrolls to the first change

### Test Case 3: Long String Wrapping
1. Create or find a secret with a very long value (e.g., a long URL or connection string)
2. Request an approval with a change to this secret
3. **Expected**:
   - The text wraps within the container instead of overflowing
   - Horizontal scrolling is only needed if absolutely necessary
   - Word-level highlighting still works correctly across wrapped lines

### Test Case 4: Auto-Scroll Behavior
1. Navigate to a secret approval request with changes
2. If the first change is not visible in the viewport
3. **Expected**:
   - The container automatically scrolls to show the first change
   - Only the inner diff container scrolls, not the entire page
   - Scrolling is smooth and centers the first change in the viewport

### Test Case 5: Visual Design
1. Navigate to a secret approval request
2. Observe the diff containers
3. **Expected**:
   - Previous version container has a darker red-tinted background (`#120808`)
   - New version container has a darker green-tinted background (`#081208`)
   - Colors provide good contrast against the card backgrounds
   - Highlighting is clearly visible

### Test Case 6: Edge Cases
1. Test with secrets that have:
   - Only whitespace changes
   - Special characters and unicode
   - Very long single words (no spaces)
   - Empty to non-empty or vice versa
2. **Expected**: All cases handle gracefully without breaking the UI

## Type

- [ ] Fix
- [x] Feature
- [ ] Improvement
- [ ] Breaking
- [ ] Docs
- [ ] Chore

## Checklist

- [x] Title follows the [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) format: `type(scope): short description` (scope is optional, e.g., `fix: prevent crash on sync` or `fix(api): handle null response`). 
- [x] Tested locally
- [ ] Updated docs (if needed)
- [x] Read the [contributing guide](https://infisical.com/docs/contributing/getting-started/overview)

### CODE REVIEW

**Comment by maidul98** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/maidul98/pr-checks/cb07f28e-9ac2-4d07-a43e-0d093c1d4dd6?refs=01984521c73ea6b52a264e218fbece3bccb88be6&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: frontend/src/pages/secret-manager/SecretApprovalsPage/components/SecretApprovalRequest/components/SecretApprovalRequestChangeItem.tsx [hunk 1]
```diff
@@ -17,6 +17,291 @@ import { twMerge } from "tailwind-merge";
...
+        while (oldIndex < nextNewMatch) {
+          result.push({
+            type: DIFF_TYPE.DELETED,
+            oldLine: oldLines[oldIndex]
+          });
+          oldIndex++;
+        }
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (IgorHorta) [reply frontend/src/pages/secret-manager/SecretApprovalsPage/components/SecretApprovalRequest/components/SecretApprovalRequestChangeItem.tsx:76]: RE2 seems to be a nodejs library .. searching for similar ones for frontend

[the remark this answers is not in the document]
Comment (IgorHorta) [reply frontend/src/pages/secret-manager/SecretApprovalsPage/components/SecretApprovalRequest/components/SecretApprovalRequestChangeItem.tsx:76]: current Regex is safe

image

**Comment by victorvhs017** [comment]
If I add a secret with whitespace at the start and the end of a secret value, it shows the change here:

image

But it doesn't show here

image

And if I save and merge, it's not there.

It would be good to standardize the behavior:
- trim and not commit anything
- do not trim and add the whitespace to the value

**Comment by victorvhs017** [comment]
When we have multiple words changed, it highlights one by one; in this case, it should be the whole block.

image

**Comment by victorvhs017** [comment]
> When we have multiple words changed, it highlights one by one; in this case, it should be the whole block.
> 
> image

This is another example 
image

What I did here is that I removed something at the end and added something at the end of the text

File: frontend/src/pages/secret-manager/SecretApprovalsPage/components/SecretApprovalRequest/components/SecretApprovalRequestChangeItem.tsx [hunk 2]
```diff
@@ -64,6 +372,47 @@ export const SecretApprovalRequestChangeItem = ({
...
+        if (!isVisible) {
+          // Scroll to center the element in the container
+          const targetScrollTop = elementTop - containerHeight / 2 + elementHeight / 2;
+          container.scrollTo({
+            top: Math.max(0, targetScrollTop),
+            behavior: "smooth"
+          });
+        }
+      }
+    };
```
Comment (victorvhs017) [inline frontend/src/pages/secret-manager/SecretApprovalsPage/components/SecretApprovalRequest/components/SecretApprovalRequestChangeItem.tsx]: nit: we can move this one to outside the component too.

File: frontend/src/pages/secret-manager/SecretApprovalsPage/components/SecretApprovalRequest/components/SecretApprovalRequestChangeItem.tsx [hunk 3]
```diff
@@ -17,6 +17,314 @@ import { twMerge } from "tailwind-merge";
...
+      } else {
+        while (oldIdx < nextNewMatch) {
+          result.push({ type: DIFF_TYPE.DELETED, text: oldWords[oldIdx] });
+          oldIdx += 1;
+        }
+      }
+    }
+  }
+  return result;
+};
```
Comment (victorvhs017) [inline frontend/src/pages/secret-manager/SecretApprovalsPage/components/SecretApprovalRequest/components/SecretApprovalRequestChangeItem.tsx]: We can move this to `lib/fn/diff.ts`, to make the file a bit smaller, and it will probably be used by the secrets overview page in the future too, when we give it the same treatment.

**Comment by IgorHorta** [comment]
> If I add a secret with whitespace at the start and the end of a secret value, it shows the change here:
> 
> image
> But it doesn't show here
> 
> image
> And if I save and merge, it's not there.
> 
> It would be good to standardize the behavior:
> 
> * trim and not commit anything
> * do not trim and add the whitespace to the value

fixed!

File: frontend/src/pages/secret-manager/SecretApprovalsPage/components/SecretApprovalRequest/components/SecretApprovalRequestChangeItem.tsx [hunk 4]
```diff
@@ -106,22 +455,65 @@ export const SecretApprovalRequestChangeItem = ({
...
+                          {renderMultilineDiffForApproval(
+                            secretVersion?.secretValue ?? "",
+                            newVersion?.secretValue ?? "",
+                            true
+                          )}
                         </div>
-                      )}
+                      </div>
+                    )
+                  ) : (
```
Comment (victorvhs017) [inline frontend/src/pages/secret-manager/SecretApprovalsPage/components/SecretApprovalRequest/components/SecretApprovalRequestChangeItem.tsx]: The nested ternaries (A ? B : C) are a bit hard to follow in this file.

We can refactor it into helper functions and components in this same file, or at least compute the conditions into variables to make it more readable, for example:

```
  const oldSecretValue = secretVersion?.secretValue ?? "";
  const newSecretValue = newVersion?.secretValue ?? "";
  const hasValueChanges = oldSecretValue !== newSecretValue;
  const isBothSingleLine = isSingleLine(oldSecretValue) && isSingleLine(newSecretValue);
```

File: backend/src/ee/services/license/license-fns.ts
Comment (victorvhs017) [inline backend/src/ee/services/license/license-fns.ts:1]: Be careful to not commit your changes to this file. We need to revert these.

File: frontend/src/pages/secret-manager/CommitDetailsPage/components/SecretVersionDiffView/SecretVersionDiffView.tsx [hunk 5]
```diff
@@ -46,47 +46,86 @@ const isArray = (obj: JsonValue): obj is JsonArray => {
...
-  if (a == null || b == null) return false;
-  if (typeof a !== typeof b) return false;
-
-  if (isArray(a) && isArray(b)) {
-    if (a.length !== b.length) return false;
-    return a.every((item: JsonValue, index: number) => deepEqual(item, b[index]));
+// Normalize secret value to match backend behavior:
+// - Trim whitespace
+// - If value ends with \n, preserve it (but trim everything else)
+// This matches the backend transform: (val.at(-1) === "\n" ? `${val.trim()}\n` : val.trim())
```
Comment (victorvhs017) [inline frontend/src/pages/secret-manager/CommitDetailsPage/components/SecretVersionDiffView/SecretVersionDiffView.tsx]: nit: I wouldn't add the actual backend code here, as it can change in the future (but keep the same behavior) and mislead other devs that may look for this exact code.

File: frontend/src/pages/secret-manager/CommitDetailsPage/components/SecretVersionDiffView/SecretVersionDiffView.tsx [hunk 6]
```diff
@@ -46,47 +46,86 @@ const isArray = (obj: JsonValue): obj is JsonArray => {
...
+// Normalize a JSON value for comparison - specifically handles secretValue strings
+const normalizeValueForComparison = (value: JsonValue, key?: string, path?: string): JsonValue => {
+  // Check if this is a secretValue field - either by key name or path
+  const isSecretValue =
+    key === "secretValue" || path?.endsWith(".secretValue") || path === "secretValue";
+
+  // Only normalize secretValue strings
+  if (isSecretValue && typeof value === "string") {
+    return normalizeSecretValue(value);
+  }
```
Comment (victorvhs017) [inline frontend/src/pages/secret-manager/CommitDetailsPage/components/SecretVersionDiffView/SecretVersionDiffView.tsx]: We should do the same for secret key, tags, and comments, at least for the whitespace trim when checking for changes.

This would go beyond the PR scope. Let's create a new issue for that in Linear.

File: frontend/src/pages/secret-manager/SecretApprovalsPage/components/SecretApprovalRequest/components/SecretApprovalRequestChangeItem.tsx (hunk 2 above)
Comment (IgorHorta) [reply frontend/src/pages/secret-manager/SecretApprovalsPage/components/SecretApprovalRequest/components/SecretApprovalRequestChangeItem.tsx]: on it!

File: frontend/src/pages/secret-manager/SecretApprovalsPage/components/SecretApprovalRequest/components/SecretApprovalRequestChangeItem.tsx (hunk 4 above)
Comment (IgorHorta) [reply frontend/src/pages/secret-manager/SecretApprovalsPage/components/SecretApprovalRequest/components/SecretApprovalRequestChangeItem.tsx]: Totally agree!

**Comment by victorvhs017** [comment]
Here, I added a new secret and there's a space splitting the first line from the rest of the value

image

File: frontend/src/pages/secret-manager/SecretApprovalsPage/components/SecretApprovalRequest/components/SecretApprovalRequestChangeItem.tsx [hunk 7]
```diff
@@ -25,6 +25,29 @@ const DIFF_TYPE = {
...
+// - If value ends with \n, preserve it (but trim everything else)
+// This matches the backend transform: (val.at(-1) === "\n" ? `${val.trim()}\n` : val.trim())
+const normalizeSecretValue = (value: string | null | undefined): string => {
+  if (!value || typeof value !== "string") return "";
+  // If value ends with \n, preserve it after trimming
+  if (value.at(-1) === "\n") {
+    return `${value.trim()}\n`;
+  }
+  return value.trim();
+};
```
Comment (victorvhs017) [inline frontend/src/pages/secret-manager/SecretApprovalsPage/components/SecretApprovalRequest/components/SecretApprovalRequestChangeItem.tsx]: This is the same function from the file above; we should extract and reuse it in both files.

**Comment by victorvhs017** [comment]
Here I'm removing the first empty line, and this is the change I'm getting:

image

image

File: frontend/src/pages/secret-manager/SecretApprovalsPage/components/SecretApprovalRequest/components/SecretApprovalRequestChangeItem.tsx (hunk 7 above)
Comment (IgorHorta) [reply frontend/src/pages/secret-manager/SecretApprovalsPage/components/SecretApprovalRequest/components/SecretApprovalRequestChangeItem.tsx]: only used here now!

**Comment by victorvhs017** [comment]
@greptileai re-review and update your summary
